Feature/first version

AnatoleMombrun opened the pull request:
# Goal of the PR
Have a first functional version of the nt user script which uses Excel for input and output, and which does the extraction fo the nt users info in parallel.

# Details
- First version of the tool working with an Excel document for input and ouput
- Excel path given in the YAML configuration file.
- Minimal Python version set to 3.8
- Parallel execution of 'net user' commands using ThreadPoolExecutor

# Not done in this PR:
- unit tests

Commit:
update readme.md

Commit:
Files from local project added : 3 modules

LoicRiegel commented on pyproject.toml:
```diff
@@ -2,12 +2,13 @@
 version = "0.1.0"
 description = ""
-authors = ["Loïc Riegel <loic.riegel@outlook.fr>"]
+authors = ["Anatole Mombrun <anatole.mombrun@gmail.com>"]
```
Please revert this change. This field should indicate the owner /main maintainer of the repo, which is me.
You can look [here](https://github.com/eclipse/kiso-testing/blob/master/pyproject.toml) (the repo of the ITF project) to see that only the owner is listed here.
Your name will be seen on GitHub since you will be the author of some commits!

LoicRiegel commented on src/nt_users_tool/constants.py:
```diff
@@ -0,0 +1,47 @@
+from lib2to3.pgen2.token import NAME
```
What is this?
Maybe this was generated by VS code and you didn't notice it?

AnatoleMombrun replied:
Indeed. Conflicting with a custom variable named "NAME". Removed.

LoicRiegel commented on src/nt_users_tool/excelprocessing.py:
```diff
@@ -0,0 +1,82 @@
+from openpyxl import Workbook, load_workbook
+from datetime import date
+
+from constants import *
```
"Wildcard imports" (= `import *`) should be banned :)
Please use `from xxx import THIS, THAT` instead (so only what you need, and this is explicit)

Pleaes make this improvement for all the files :)

AnatoleMombrun replied:
Will be changed.

LoicRiegel replied:
I see you removed them, thanks

LoicRiegel commented on src/nt_users_tool/netcommands.py:
```diff
@@ -0,0 +1,31 @@
+        dict: The dictionnary containing relevant information about this user
+    """
+    input = nt_user_id.upper()
+    command = 'net user /domain ' + input
```
use f-string here?

AnatoleMombrun replied:
Syntax changed to f-string.

LoicRiegel commented on src/nt_users_tool/netcommands.py:
```diff
@@ -0,0 +1,31 @@
+from constants import *
+
+
+def obtain_info_from_user(nt_user_id: str):
```
This function does too many things in my opinion.

What do you think about having those function?
```Python
def get_nt_user_info(nt_user:string) -> str
    """Returns the nt user info for the given NT user as a string by executing the 'net command /domain' command."""

def get_all_nt_users_info(nt_users:list[str]) -> list[str]:
    """Returns all nt users info in a list."""

def extract_nt_user_info(nt_user_info:str) -> NTUserInfo:
    """Extract NTUser info from a string and returns it as a NTUserInfo object.""" # NTUserInfo can be a typing.NamedTuple for instance

def extract_all_nt_user_info(nt_users_info:list[str]) -> list[NTUserInfo]:
    """Extract NTUser info from a lsit of nt users string and returns it as a list of NTUserInfo object."""
```

AnatoleMombrun replied:
Changes made accordingly to suggestion.

LoicRiegel commented on src/nt_users_tool/netcommands.py:
```diff
@@ -0,0 +1,31 @@
+        string_input (str): the nt_user_id for which the command will be executed
+
+    Returns:
+        dict: The dictionnary containing relevant information about this user
```
A dictionary would work to store the relevant information about a user, but it's nice to be even more precise when defining types to clearly indicate what we expect. What are those relevant information about the user?

In my opinion, you can use a `NamedTuple` (in the typing module), simple and enough to do the job :)

AnatoleMombrun replied:
Data format changed from dict to NamedTuple subclass NTUserInfo.

LoicRiegel commented on src/nt_users_tool/yamlprocessing.py:
```diff
@@ -0,0 +1,31 @@
+import yaml
+
+
+def output_file_path(config_file_name: str):
```
Currently, you open the file and read from it twice, which is not optimal. Here is what I suggest:
* have 1 function to read from the file and just return the content
* have 1 function to get a particular config from the content. Inside this function, we should handle the possibility that the config that we are currently trying to get is missing! In this case, let's throw an exception and exit the script right away.

What do you think?

AnatoleMombrun replied:
The current functions will be removed. Your solution is adopted.

LoicRiegel commented on src/nt_users_tool/excelprocessing.py:
```diff
@@ -0,0 +1,82 @@
+    :param user_info: _description_
+    """
+
+    keys = list(user_info.keys())
```
`dict.keys()` already returns a list

AnatoleMombrun replied:
It does not exactly return a list (class dict_keys). I needed to access indexes hence the typecast.

LoicRiegel commented on src/nt_users_tool/excelprocessing.py:
```diff
@@ -0,0 +1,82 @@
+    Returns:
+        list: List of nt_user IDs
+    """
+    workbook = load_workbook(excel_file_name, read_only=True)
```
I suggest to pass a workbook object to this function and do the `load_workbook` and `wb.close()` etc... in another function or in the main function.

AnatoleMombrun replied:
It will be done in the main() function.

LoicRiegel commented on src/nt_users_tool/excelprocessing.py:
```diff
@@ -0,0 +1,82 @@
+        list: List of nt_user IDs
+    """
+    workbook = load_workbook(excel_file_name, read_only=True)
+    worksheet = workbook.active
```
Here, you are assuming that when you open the Excel file for the first time, the active worksheet will always be the worksheet where the input of nt users are. I don't think that it assumption will be true all the time :) I would switch to a given worksheet, which name is given as config in the yaml file

AnatoleMombrun replied:
We no longer use excel path as an imput for this function. The worksheet name will be located in the config file.

LoicRiegel commented on src/nt_users_tool/constants.py:
```diff
@@ -0,0 +1,47 @@
+from lib2to3.pgen2.token import NAME
+
+
+BLANK = " "
```
Are you sure you need those constants?

AnatoleMombrun replied:
Was added for "cleanliness" of the code but it is unnecessary. Will be removed.

LoicRiegel commented on src/nt_users_tool/constants.py:
```diff
@@ -0,0 +1,47 @@
+BLANK = " "
+SLASH = "/"
+
+USERNAME = "user_name"
```
okay, I think I understand those constants, let's see if we can do something better :) Let's discuss it

LoicRiegel commented on src/nt_users_tool/constants.py:
```diff
@@ -0,0 +1,47 @@
+
+COLUMNS_LIST = [NT_USER_COLUMN, NAME_COLUMN, EXPIRATION_DATE_COLUMN]
+
+FAKE_NET_COMMAND = "The request will be processed at a domain controller for domain de.bosch.com\
```
those fake strings, you might be able to move them in the unit tests part when you're doing them. They might be useful for mocking

AnatoleMombrun replied:
Still not moved with commit 1ea2511, will move them later on.

AnatoleMombrun replied:
Moved in ef87cb4.

LoicRiegel reviewed (changes requested):
Looks good, let's discuss in a meeting how we can improve things :)

LoicRiegel commented on src/nt_users_tool/yamlprocessing.py:
```diff
@@ -0,0 +1,31 @@
+import yaml
+
```
Something more generic for the name of the module? "read_config.py" for instance?

BTW modules should follow the snake_case format. Can you apply that evverywhere please?

AnatoleMombrun replied:
File names will be changed.

LoicRiegel replied:
I see it's done, thanks

LoicRiegel commented on src/nt_users_tool/excelprocessing.py:
```diff
@@ -0,0 +1,82 @@
+        worksheet[column + str(row_number)] = user_info[keys[i]]
+
+
+def nt_user_list(excel_file_name: str):
```
`read_nt_users' ?

AnatoleMombrun replied:
Function name changed to read_nt_users for clarity.

AnatoleMombrun committed:
Renamed all files.
Added configuration files.
Made changes accordingly to comments in pull request.

AnatoleMombrun committed:
Changes to .toml file:
Name is now the owner of the repo
New syntax for group dependencies

LoicRiegel commented on config.yml:
```diff
@@ -0,0 +1,2 @@
+input_file_path: ../input.xlsx
```
Don't forget that the name of the input nt user sheet must also be given here :)

Also, I think that in practice, the `input_file_path` and `output_file_path` will be the same. But we can keep 2 separate config settings.

LoicRiegel commented on pyproject.toml:
```diff
@@ -1,13 +1,14 @@
-[tool.poetry]
++[tool.poetry]
```
⚠️ 
```suggestion
[tool.poetry]
```

AnatoleMombrun replied:
I also noticed it. It will be changed as the project no longer runs with this typo.

LoicRiegel commented on pyproject.toml:
```diff
@@ -1,13 +1,14 @@
 version = "0.1.0"
 description = ""
-authors = ["Loïc Riegel <loic.riegel@outlook.fr>"]
+authors = ["Loïc Rigel <loic.riegel@outlook.fr>"]
```
⚠️ 
```suggestion
authors = ["Loïc Riegel <loic.riegel@outlook.fr>"]
```

AnatoleMombrun replied:
Oops...

LoicRiegel commented on src/nt_users_tool/excel_processing.py:
```diff
@@ -0,0 +1,82 @@
+        sheets_names_list (list): Names of the sheets inside the file
+    """
+    workbook = Workbook()
+    workbook.remove(workbook.active)
```
You might not want to do that! In case the output file is the same as the input file, and you just want to add /overwrite the ouptut worksheets

AnatoleMombrun replied:
Okay. This was done in case the output file was not the same as the input.

LoicRiegel commented on src/nt_users_tool/net_commands.py:
```diff
@@ -0,0 +1,66 @@
+from typing import NamedTuple
+from constants import LAST_RELEVANT_ELEMENT_POSITION
+
+class NTUserInfo(NamedTuple):
```
You can also add a docstring for the class. For is this data structure for? What data does it hold?

AnatoleMombrun replied:
Okay. Will add.

LoicRiegel commented on src/nt_users_tool/read_config.py:
```diff
@@ -0,0 +1,21 @@
+import yaml
+from constants import CONFIG_FILE_PATH, INPUT_FILE, OUTPUT_FILE
+
+def read_config_file() -> dict():
```
⚠️ do not create a new dictionary here!
```suggestion
def read_config_file() -> dict:
```

If you want to be more precise, you can also specify the type of the dict's keys and values. For instance:
```Python
from typing import Dict, Any
def read_config_file() -> Dict[str, Any]:
```
(`dict[str, Any]` works with Python >= 3.10)

AnatoleMombrun replied:
Typo. I tried to apply the syntax you described but does not work with python 3.8.5 as you mentionned.

AnatoleMombrun replied:
Now fixed in 638ef90.

LoicRiegel commented on src/nt_users_tool/excel_processing.py:
```diff
@@ -0,0 +1,82 @@
+from openpyxl import Workbook, load_workbook
+from datetime import date
+
+from constants import *
```
Looks like you forgot this  `import *`  :)

AnatoleMombrun replied:
fixed in 04b3eac

AnatoleMombrun committed:
Changed the contents of excel_processing.
Fixed typo in toml file.

AnatoleMombrun committed:
New changes to excel_processing.py
Added documentation for feel_all_sheets because of new implementation
Changed create_result_sheets function

LoicRiegel commented:
I switched the destination branch to `develop`

LoicRiegel commented on src/nt_users_tool/net_commands.py:
```diff
@@ -0,0 +1,72 @@
+    :return: The response as a string.
+    """
+    input = nt_user.upper()
+    command = (f"net user /domain {input}")
```
there is no need for brackets
```suggestion
    command = f"net user /domain {input}"
```

AnatoleMombrun replied:
Brackets removed.

LoicRiegel commented on src/nt_users_tool/net_commands.py:
```diff
@@ -0,0 +1,72 @@
+
+from typing import NamedTuple
+
+from constants import LAST_RELEVANT_ELEMENT_POSITION
```
Please use relative imports like this `from .constants ...` since we have a package. Can you please make the change in all imports?

We could also import from the package `from nt_user_tool.constants import ...`

AnatoleMombrun replied:
Imports are now done with the package syntax nt_user_tool.module.

LoicRiegel commented on config.yml:
```diff
@@ -0,0 +1,2 @@
+input_file_path: ./input.xlsx
```
let's name this file with ".yaml" instead

AnatoleMombrun replied:
Fixed in 04d8101.

LoicRiegel commented on src/nt_users_tool/excel_processing.py:
```diff
@@ -0,0 +1,82 @@
+        worksheet[column + str(row_number)] = nt_user_info[i]
+
+
+def fill_all_sheets(workbook: Workbook, nt_user_info_list: list):
```
here you can be more precise about the type hint
```Python
nt_user_info_list: list[NTUserInfo]
```
It helps the IDE understand what you are passing to the function which can then help you debug way more easilly :)

AnatoleMombrun replied:
![image](https://user-images.githubusercontent.com/98102969/197147759-4b47a1aa-37be-4d92-b920-e9a2e46f280a.png)

AnatoleMombrun replied:
Fixed in 638ef90.

LoicRiegel replied:
yes, you can only do that in Python 3.10
With older versions you have to do
```Python
from typing import List
List[MyElement]
```

LoicRiegel commented on src/nt_users_tool/excel_processing.py:
```diff
@@ -0,0 +1,82 @@
+        fill_one_row(users_sheet, all_user_index, COLUMNS_LIST, user)
+        all_user_index += 1
+
+        user_day, user_month, user_year = user.expiration_date.split("/")
```
Do you need to create those variables?
Could you use `user.expiration_date.day`, `user.expiration_date.month` etc... instaed?

AnatoleMombrun replied:
Those variables are created because when fetching from the net command, we obtain a date that looks like DD/MM/YYYY. I don't know how to write the syntax with expiration_date.month or .day. Could you please support me on that?

AnatoleMombrun replied:
Fixed in 04d8101.

LoicRiegel commented on src/nt_users_tool/excel_processing.py:
```diff
@@ -0,0 +1,82 @@
+        user_day, user_month, user_year = user.expiration_date.split("/")
+        user_month = int(user_month)
+        user_year = int(user_year)
+        if user_year < now_year:
```
This logic is there to determine if a user is expired or not, or if he/she is expiring soon. This logic thus don't belong here in the excel processing file.
I suggest creating a new module called "nt_user.py" or "nt_user_info.py". Inside it, you will put the NTUserInfo class and some functions to compute if the user is expired or not.
I would recommand to use an enum for the expiration status (okay, expires soon, or expired).

In the top of the module, you can craete an object `date_now = date.now()`. This object will only be generated once, when python imports this module. You can then use this object inside the module. I am suggesting this so that we don't have to pass the current date as an argument to this function to determinte the expiration status. Let me know if what I 'm saying is clear aha :)

AnatoleMombrun replied:
Fixed in 04d8101.

LoicRiegel commented on src/nt_users_tool/excel_processing.py:
```diff
@@ -0,0 +1,82 @@
+                expiring_index +=1
+        elif user_year - now_year < TWO_YEAR_GAP:
+            if now_month > user_month:
+                if (user_month - now_month)%12 <= MONTHS_TO_EXPIRE:
```
Maybe you can simplify this logic. datetime.date objects can be compared easily
```Python
from datetime import date

d_after = date(2023, 2, 1)
d_before = date(2022, 12, 1)

elapsed = d_after - d_before
print("Nb days:", elapsed.days) # Will print 62
```

AnatoleMombrun replied:
Fixed in 04d8101.

LoicRiegel commented on src/nt_users_tool/net_commands.py:
```diff
@@ -0,0 +1,72 @@
+    """
+    full_name: str
+    nt_user: str
+    expiration_date: str
```
I suggest to use a more appropriate type for that, maybe datetime.date

LoicRiegel replied:
great!

LoicRiegel commented on src/nt_users_tool/net_commands.py:
```diff
@@ -0,0 +1,72 @@
+def extract_nt_user_info(nt_user_string: str) -> NTUserInfo:
+    """Extracts relevant data from a string generated with a given nt_user.
+
+    :param nt_user_string: Net command response as a string with given nt_user. 
```
Based on the description of this param, I would recommand naming it something like "nt_command_response"

AnatoleMombrun replied:
Fixed in 04d8101.

LoicRiegel reviewed (commented):
Thank you very much for fixing lots of things already! I put more comments, have fun 😂

AnatoleMombrun committed:
Changes according to recent comment.
Added nt_user_info module.
Changed NTUserInfo dataclass attribute to datetime.date
Changed logic to accomodate the new attribute.
Fixed errors and typos.

AnatoleMombrun committed:
Updated cli.py and added threadpoolexecutor.

AnatoleMombrun committed:
changed constants.py file and fixed ThreadPoolExecutor missing.

AnatoleMombrun committed:
Added new type hints and fixed typos.

AnatoleMombrun committed:
Improvements to cli and behavior changes.
Added try except blocks to cli.py.
Removed useless print in net_commands and nt_user_info.
Changed behavior of excel_processing.create_result_sheets.

AnatoleMombrun committed:
Multithreading implementation in net_commands and column width fix in excel_processing.

AnatoleMombrun committed:
Changed behavior of excel_processing.
User expiration date are now seperated in 3 categories.
Constants have been added to reflect this change.
Columns are now deleted in the sheet that are already present if using load_workbook.

AnatoleMombrun committed:
Main docstring and table created for all_users_sheet

AnatoleMombrun committed:
Fixed wrong date calculation in nt_user_info.
Fixed table management (now deleted if already present).

LoicRiegel committed:
remove unused import

LoicRiegel committed:
quick fixes read_config.py

LoicRiegel committed:
quick fixes excel_processing.py

LoicRiegel committed:
quick fixes excel_processing.py continue

LoicRiegel committed:
Add security check to prevent code injection + quick improvements in cli.py

LoicRiegel committed:
Remove sensible info from examples and docstrings

LoicRiegel committed:
created nt_user_check.py

LoicRiegel committed:
Security check in function and remove nt_user_check.py

LoicRiegel committed:
improvements src/nt_users_tool/net_commands.py

LoicRiegel commented:
@AnatoleMombrun I just pushed new commits. You can have a look. I also updated the PR description.
Those are mainly minor fixes.

I also added a simple security check before doing the popen command to avoid code injection attacks.

Also, this project is public, so keep in mind that no sensible information should be pushed here.


Let's test the new commits tomorrow, and if the script works, we merge this PR :)

AnatoleMombrun committed:
toml filed changed to select pypi source for packages.

AnatoleMombrun committed:
Typo in nt_user_info: unecessary semicolon

AnatoleMombrun committed:
Added comma to last import item in nt_user_info

LoicRiegel reviewed (approved):
LGTM now, thank you a lot for this work!
